### PR TITLE
Add WeightDropout

### PR DESCRIPTION
**Pull request by scottire**
Add WeightDropout from [fastai](https://github.com/fastai/fastai/blob/release-1.0.55/fastai/text/models/awd_lstm.py#L29), which is inspired by [Regularizing and Optimizing LSTM Language Models](https://arxiv.org/abs/1708.02182), (and [awd-lstm-lm](https://github.com/salesforce/awd-lstm-lm/blob/master/weight_drop.py)). 

This is my first pull request for AllenNLP so I thought I should open this pull request early to get feedback. Happy to write tests for this and fix any issues.

To get it working, it uses the PyTorch RNN `_module` directly to update the weights and then uses the `PytorchSeq2SeqWrapper` for the forward pass. The linter is not happy about that so maybe there is a better solution that I haven't thought about.

### CODE REVIEW

**Comment by matt-gardner** [comment]
Hey, thanks for the contribution.  Just FYI, we have an internal hackathon for the rest of this week, so we likely won't get to this until next week.

**Comment by matt-gardner** [comment]
@rloganiv, I know you're in Italy for ACL, but I wanted to check with you about this - did you end up implementing something like this when you re-implemented AWD-LSTM?

**Comment by scottire** [comment]
Great, thanks for the review. Will update when I get a chance.

File: allennlp/modules/weight_drop.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            #Makes a copy of the weights of the selected layers.
+            w = getattr(self.module, layer)
+            self.register_parameter(f'{layer}_raw', nn.Parameter(w.data))
+            self.module._parameters[layer] = F.dropout(w, p=self.weight_p, training=False)
+
+    def _setweights(self):
+        "Apply dropout to the raw weights."
+        for layer in self.layer_names:
+            raw_w = getattr(self, f'{layer}_raw')
+            self.module._parameters[layer] = F.dropout(raw_w, p=self.weight_p, training=self.training)
```
[the remark this answers is not in the document]
Comment (rloganiv) [reply allennlp/modules/weight_drop.py]: Nevermind - I see now that this is necessary to avoid raising a `ValueError`.

**Comment by scottire** [comment]
I have merged [this pull request](https://github.com/scottire/allennlp/pull/1) from @rloganiv which fixes all of the issues that are mentioned above.

**Comment by matt-gardner** [comment]
@scottire, @rloganiv, is this ready to be merged?

**Comment by rloganiv** [comment]
Yep

**Comment by matt-gardner** [comment]
Thanks @scottire and @rloganiv!  Rob, I'm assuming this is largely intended for language modeling uses, and should end up in the new `allennlp-language-models` repo that you're setting up, once we split things out.

**Pull request merged**